### PR TITLE
boost for first letter matches

### PR DESCRIPTION
**Pull request by willmcgugan**
Tweak to the command palette matcher to boost matches for the first word(s).

Consider the system commands. If I were to type "SS" I would get "**S**how Key**s** and Help Panel". Now I get "**S**ave **S**creenshot" because I'm matching the first letters.

Knowing this makes it easier to get a precise match.

### CODE REVIEW

File: src/textual/fuzzy.py [hunk 1]
```diff
@@ -40,6 +40,10 @@ def __init__(
             ".*?".join(f"({escape(character)})" for character in query),
             flags=0 if case_sensitive else IGNORECASE,
         )
+        self._first_word_regex = compile(
```
Comment (darrenburns) [inline src/textual/fuzzy.py:43]: Is a new `Matcher` created on each key press in the command palette？

Comment (willmcgugan) [reply src/textual/fuzzy.py:43]: Yeah

**Comment by willmcgugan** [comment]
Actually, I've nerd sniped myself. We can do so much better with the matching, but regexes aren't going to be enough.

**Comment by willmcgugan** [comment]
Putting this on hold. I'm going to show great strength of character and not work on this, because there are higher priorities right now.

**Comment by darrenburns** [comment]
@willmcgugan Do you want to merge this to get the improvement in or stick with what we have?

**Comment by willmcgugan** [comment]
Going to hold off merging this. It's not quite a 100% improvement.

**Comment by willmcgugan** [comment]
@darrenburns slept on it, and came up with this.

You may want to check it against Posting.

**Comment by darrenburns** [comment]
Cool - I'll try it against Posting soon.

**Comment by darrenburns** [comment]
Minor but these look the wrong way around:

image

Should the first `s` in `response` not match here, and mean that these results should be in the opposite order?

Also I'm not sure about underlining the spaces.

**Comment by willmcgugan** [comment]
Wait, I see what you mean. It could match `res` and `sec`.

I may be able to tweak the heuristic. ATM it gives a boost to single words. It could boost based on the least number of groups.

**Comment by willmcgugan** [comment]
@darrenburns tweaked the heuristic so it favors matches with fewer groups.

**Comment by darrenburns** [comment]
I'm not sure that was the problem as in the original example wasn't there 4 matching groups in both results?

It still seems like the first `s` in `response` is being skipped here, and the results aren't what I'd expect.

image

**Comment by willmcgugan** [comment]
My estimate for a reasonable maximum on the number of loops was too low.

Seems to work well with your example now. "exp res" puts response at the top.

**Review by darrenburns — APPROVED** [review]

**Pull request merged**